core/src/main/java/io/substrait/relation/Expand.java

```diff
@@ -0,0 +1,64 @@
+package io.substrait.relation;
+
+import io.substrait.expression.Expression;
+import io.substrait.type.Type;
+import io.substrait.type.TypeCreator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.immutables.value.Value;
+
+@Value.Enclosing
+@Value.Immutable
+public abstract class Expand extends SingleInputRel {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(Expand.class);
+
+  public abstract List<ExpandField> getFields();
+
+  @Override
+  public Type.Struct deriveRecordType() {
+    Type.Struct initial = getInput().getRecordType();
+    return TypeCreator.of(initial.nullable())
+        .struct(
+            Stream.concat(
+                initial.fields().stream(),
+                getFields().stream()
+                    .map(
+                        f -> {
+                          if (f.getSwitchingField().isPresent()) {
+                            return f.getSwitchingField().get().getDuplicates().get(0).getType();
+                          } else {
+                            return f.getConsistentField().get().getType();
+                          }
+                        })));
+  }
+
+  @Override
+  public <O, E extends Exception> O accept(RelVisitor<O, E> visitor) throws E {
+    return visitor.visit(this);
+  }
+
+  public static ImmutableExpand.Builder builder() {
+    return ImmutableExpand.builder();
+  }
+
+  @Value.Immutable
+  public abstract static class ExpandField {
+    public abstract Optional<SwitchingField> getSwitchingField();
+
+    public abstract Optional<Expression> getConsistentField();
+
+    public static ImmutableExpand.ExpandField.Builder builder() {
+      return ImmutableExpand.ExpandField.builder();
+    }
+  }
```

Comment:
A class with two optional fields, only 1 of which is set at any point in time, feels like a code smell to me. I think this should be an interface `ExpandField` with two implementing classes.